src/common.rs

```diff
@@ -2,6 +2,29 @@
 
 use arrayvec::ArrayVec;
 
+/// Find real roots of a polynomial up to order of 3.
+/// Returns values of x for which c0 + c1 x + c2 x² + c3 x³ = 0, allowing for the coefficients
+/// to be equal to 0.
+pub(crate) fn solve_poly_3(c0: f64, c1: f64, c2: f64, c3: f64) -> ArrayVec<[f64; 3]> {
```

Comment:
Generally this looks good to me, but I'm still not sure it's as numerically robust as possible. Coefficients of exactly 0 are handled well here, but extremely tiny coefficients still are not. There's a TODO in `solve_quadratic` for this.